services/dynamic/dynamic-regex.service.js

```diff
@@ -0,0 +1,99 @@
+import Joi from 'joi'
+import RE2 from 're2'
+import { BaseService, InvalidParameter, queryParams } from '../index.js'
+import { url } from '../validators.js'
+import { renderDynamicBadge } from '../dynamic-common.js'
+
+export default class DynamicRegex extends BaseService {
+  static category = 'dynamic'
+  static route = {
+    base: `badge/dynamic/regex`,
+    pattern: '',
+    queryParamSchema: Joi.object({
+      url,
+      search: Joi.string().required(),
+      replace: Joi.string().optional(),
+      flags: Joi.string().optional(),
+      noMatch: Joi.string().optional(),
+    }),
+  }
+  static openApi = {
+    '/badge/dynamic/regex': {
+      get: {
+        summary: 'Dynamic Regex (re2) Badge [Experimental]',
+        description:
+          '⚠️ Experimental: This badge is considered experimental and may change or be removed at any time.\n\nThis badge will extract text from a file using re2 (a subset of regex: https://github.com/google/re2).\nThe main use-case is to extract values from unsupported plain-text files.\nFor example: if a file contains a line like `version - 2.4` you can extract the value `2.4` by using a search regex of `version - (.*)` and `$1` as replacement.\n\nFull Syntax documentation here: https://github.com/google/re2/wiki/Syntax',
+        parameters: queryParams(
+          {
+            name: 'url',
+            description:
+              'The URL to a file to search. The full raw content will be used as the search string.',
+            required: true,
+            example:
+              'https://raw.githubusercontent.com/badges/shields/refs/heads/master/README.md',
+          },
+          {
+            name: 'search',
+            description:
+              'A re2 expression that will be used to extract data from the document. Only the first matched text will be returned.',
+            required: true,
+            example: 'Every month it serves (.*?) images',
+          },
+          {
+            name: 'replace',
+            description:
+              'A regex expression that will be used as the replacement of the search regex, like `$1` to specify the first matched group, etc. If empty (default), no replacement will be done and the full matched text will be shown.',
```

Comment:
Is "regex expression" the right terminology here? AFAIK It is a replacement string that can contain patterns.
It might be useful to show examples of a positional and named capture group here e.g:

- `Every (.*?) it serves (.*?) images` and `$2`
- `Every month it serves (?<num>.*?) images` and `$<num>`

just to hint at the possibilities a bit more.

Reply:
I wasn't aware of that!
I tried to find documentation about the replacement syntax, but I didn't find anything. Do you know if I can link to documentation somewhere or did you just find that by testing?

Reply:
I looked at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_the_replacement
but I'm not sure that is the best thing to link to as ``$` `` and `$'` aren't relevant in this context.